### PR TITLE
fix: not forward admin API

### PR DESCRIPTION
**Pull request by cindyyan317**
Don't forward admin API to rippled

### CODE REVIEW

File: src/rpc/RPCHelpers.cpp [hunk 1]
```diff
@@ -1273,6 +1273,23 @@ specifiesCurrentOrClosedLedger(boost::json::object const& request)
...
+    auto const isFieldSet = [&request](auto const field) {
+        return request.contains(field) and request.at(field).is_bool() and request.at(field).as_bool();
+    };
+    if (method == JS(ledger)) {
+        if (isFieldSet(JS(full)) or isFieldSet(JS(accounts)) or isFieldSet(JS(type))) {
+            return true;
+        }
+    }
+
+    if (method == JS(feature) and request.contains(JS(vetoed)))
```
Comment (cindyyan317) [inline src/rpc/RPCHelpers.cpp:1288]: Should we allow forwarding if vetoed is false?

Comment (godexsoft) [reply src/rpc/RPCHelpers.cpp:1288]: Absolutely not. Clio should never be connecting to validator nodes in the first place.

File: src/rpc/RPCHelpers.cpp [hunk 2]
```diff
@@ -1273,6 +1273,23 @@ specifiesCurrentOrClosedLedger(boost::json::object const& request)
     return false;
 }
 
+bool
+isAdminCmd(std::string const& method, boost::json::object const& request)
+{
+    auto const isFieldSet = [&request](auto const field) {
+        return request.contains(field) and request.at(field).is_bool() and request.at(field).as_bool();
+    };
```
Comment (godexsoft) [inline src/rpc/RPCHelpers.cpp:1281]: style nit: Let's separate the lambda from its usage by a whitespace line.

File: src/rpc/RPCEngine.hpp [hunk 3]
```diff
@@ -131,8 +132,13 @@ class RPCEngine {
     Result
     buildResponse(web::Context const& ctx)
     {
-        if (forwardingProxy_.shouldForward(ctx))
+        if (forwardingProxy_.shouldForward(ctx)) {
+            // Disallow forwarding of the admin api, only user api is allowed for security reasons.
+            if (isAdminCmd(ctx.method, ctx.params)) {
```
Comment (godexsoft) [inline src/rpc/RPCEngine.hpp]: nit: Would be ok to remove the curlies imo.

File: src/rpc/RPCHelpers.cpp [hunk 4]
```diff
@@ -1273,6 +1273,23 @@ specifiesCurrentOrClosedLedger(boost::json::object const& request)
...
+    if (method == JS(ledger)) {
+        if (isFieldSet(JS(full)) or isFieldSet(JS(accounts)) or isFieldSet(JS(type))) {
```
Comment (godexsoft) [inline src/rpc/RPCHelpers.cpp]: nit: Curlies not needed imo

File: src/rpc/RPCHelpers.hpp [hunk 5]
```diff
@@ -557,6 +557,16 @@ parseIssue(boost::json::object const& issue);
 bool
 specifiesCurrentOrClosedLedger(boost::json::object const& request);
 
+/**
+ * @brief Check if a request requires ADMIN role to execute in rippled.
```
Comment (godexsoft) [inline src/rpc/RPCHelpers.hpp]: Maybe something like "Check whether a request is considered to require administrative privileges on `rippled` side" is more accurate? In reality some APIs require IDENTIFIED role, not ADMIN on `rippled` side.

**Review by godexsoft — CHANGES_REQUESTED** [review]

File: src/rpc/RPCHelpers.cpp (hunk 1 above)
Comment (cindyyan317) [reply src/rpc/RPCHelpers.cpp:1288]: I see. The `vetoed` is to really vote on something.

**Review by godexsoft — DISMISSED** [review]
LGTM 👍

File: tests/unit/rpc/RPCHelpersTests.cpp [hunk 6]
```diff
@@ -539,3 +539,42 @@ TEST_F(RPCHelpersTest, ParseIssue)
         std::runtime_error
     );
 }
+
+TEST_F(RPCHelpersTest, isAdminCmd)
```
Comment (kuznetsss) [inline tests/unit/rpc/RPCHelpersTests.cpp]: Maybe refactor this to a parameterized tests?

**Review by kuznetsss — APPROVED** [review]
👍

**Pull request merged**